### PR TITLE
Support orc format for native batch ingestion

### PR DESCRIPTION
**Pull request by jihoonson**
### Description

Part of https://github.com/apache/incubator-druid/issues/8812. Tested reading from local, s3, and hdfs. An example config would be 

```json
    "ioConfig" : {
      "type" : "index_parallel",
      "inputSource" : {
        "type" : "s3",
        "uris" : [ "s3://path/to/your/orc/file" ]
      },
      "inputFormat" : {
        "type" : "orc"
      }
    },
```

This PR doesn't include the documentation.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/incubator-druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/incubator-druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

---
This change is [Reviewable](https://reviewable.io/reviews/apache/incubator-druid/8950)

### CODE REVIEW

File: extensions-core/orc-extensions/src/test/java/org/apache/druid/data/input/orc/OrcHadoopInputRowParserTest.java [hunk 1]
```diff
@@ -217,54 +217,70 @@ private static HadoopDruidIndexerConfig loadHadoopDruidIndexerConfig(String conf
...
+      }
+    }
+    finally {
+      reader.close();
     }
   }
 
-  private static List<InputRow> getAllRows(HadoopDruidIndexerConfig config)
-      throws IOException, InterruptedException
+  // This method returns a list of the final rows. This is definitely a bug, but it should be better to delete
```
Comment (clintropolis) [inline extensions-core/orc-extensions/src/test/java/org/apache/druid/data/input/orc/OrcHadoopInputRowParserTest.java]: thanks for fixing this, I noticed a while ago but then forgot about it :rofl:

File: extensions-core/parquet-extensions/pom.xml [hunk 2]
```diff
@@ -288,72 +288,70 @@
...
           <groupId>log4j</groupId>
+          <artifactId>log4j</artifactId>
         </exclusion>
         <exclusion>
-          <artifactId>jetty-sslengine</artifactId>
           <groupId>org.mortbay.jetty</groupId>
+          <artifactId>jetty-sslengine</artifactId>
         </exclusion>
         <exclusion>
-          <artifactId>jetty-util</artifactId>
```
Comment (clintropolis) [inline extensions-core/parquet-extensions/pom.xml]: thanks for fixing the ordering on this, however it looks like a couple of `artifactId` lines got lost

File: extensions-core/orc-extensions/src/main/java/org/apache/druid/data/input/orc/OrcReader.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      OrcStruct value = null;
+
+      @Override
+      public boolean hasNext()
+      {
+        if (value == null) {
+          try {
+            // The returned OrcStruct in next() can be kept in memory for a while.
+            // Here, we create a new instance of OrcStruct before calling RecordReader.next(),
+            // so that we can avoid to share the same reference to the "value" across rows.
```
Comment (clintropolis) [inline extensions-core/orc-extensions/src/main/java/org/apache/druid/data/input/orc/OrcReader.java]: Thanks for adding a comment

File: extensions-core/orc-extensions/src/test/java/org/apache/druid/data/input/orc/OrcHadoopInputRowParserTest.java (hunk 1 above)
Comment (jihoonson) [reply extensions-core/orc-extensions/src/test/java/org/apache/druid/data/input/orc/OrcHadoopInputRowParserTest.java]: Oh I actually fixed this method but forgot to remove the comment. Will remove it.

File: extensions-core/parquet-extensions/pom.xml (hunk 2 above)
Comment (jihoonson) [reply extensions-core/parquet-extensions/pom.xml]: Oops, fixed. Thanks!

File: core/src/main/java/org/apache/druid/data/input/IntermediateRowParsingReader.java [hunk 4]
```diff
@@ -39,25 +40,57 @@
...
+          final T row = intermediateRowIterator.next();
+          try {
+            rows = parseInputRows(row).iterator();
+          }
+          catch (IOException e) {
+            throw new ParseException(e, "Unable to parse row [%s]", row);
+          }
+        }
+
+        return rows.hasNext() || intermediateRowIterator.hasNext();
```
Comment (clintropolis) [inline core/src/main/java/org/apache/druid/data/input/IntermediateRowParsingReader.java]: I think rows will always have a next at this point in the code

Comment (jihoonson) [reply core/src/main/java/org/apache/druid/data/input/IntermediateRowParsingReader.java]: Good point. Changed to return true here.

**Comment by jihoonson** [comment]
This CI is failing because `hadoop-hdfs-client` is a runtime dependency. I'm testing it.

**Comment by jihoonson** [comment]
Ready for review.

**Comment by jihoonson** [comment]
@clintropolis thanks for the review.

**Pull request merged**